### PR TITLE
Selectively trim and collapse inter-node space

### PR DESCRIPTION
**Pull request by kemitchell**
This pull request makes changes to the algorithm for trimming and collapsing space in text nodes.

The goals of the changes are:

1.  Don't mess with space in `<pre>` tags.
2.  Try to better honor how HTML and CSS present space between contiguous, inline text- and non-text elements.
3.  Keep removing the most obviously pointless, space-only text nodes around children elements.

On point number two, an example:

```javascript
bel`
<p>
  To go back to the homepage
  <a href=/>
    click here
  </a>.
</p>
`
```

... currently produces ...

```html
<p>To go back to the homepage<a href=/>click here</a>.</p>
```

Browsers display no space between "homepage" and "click here".

I believe most folks would expect the "natural-looking" HTML template to display as if were pasted directly into HTML, equivalent to:

```html
<p>To go back to the homepage <a href=/>click here</a>.</p>
```

There are various workarounds.  All the ones I've found look regrettably hackish.

```javascript
bel`
<p>
  To go back to the homepage <a href=/>
    click here
  </a>.
</p>
`
```

```javascript
bel`
<p>
  ${document.createTextNode('To go back to the homepage ')}
  <a href=/>
    click here
  </a>.
</p>
`
```

I'm guessing my good friend @yoshuawuyts has it in for all my beautiful whitespace for very valid and impressive performance reasons.  I suppose the situation boils down to something like:

```html
<p><span>some</span> beautiful <code>code</code></p>
```

... is faster to allocate, diff, render, &c. than ...

```html
<p> <span>some</span> beautiful <code>code</code> </p>
```

... due to the pointless text nodes before the `<span>` and after the `<code>`.

This patch tries to let template authors have their whitespace, and let the rendering algos eat it, too.  Essentially, I've tried to write a loose version of the HTML/CSS inter-element space logic into the bel code that reduces whitespace.

The hard call is space between non-text elements:

```javascript
bel`
<p>
  <strong>whitespace</strong>
  <em>is beautiful</em>.
</p>
`
```

Unless we want the algorithm to start looking at the kinds of child elements (or parents) to start making educated guesses about when to drop whitespace, and when to keep it, making this work as expected means a lot of pointless text nodes for non-inline content.  If the perf gain is really that impressive, perhaps it would be worth exposing two template functions, one that munches all the whitespace it can, like XML's `collapse`, and one that leaves it alone, for the HTML/CSS renderer to figure out.

### CODE REVIEW

File: test/elements.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ test('svg', function (t) {
   </svg>`
   t.equal(result.tagName, 'svg', 'create svg tag')
   t.equal(result.childNodes[0].tagName, 'rect', 'created child rect tag')
-  t.equal(result.childNodes[1].getAttribute('xlink:href'), '#test', 'created child use tag with xlink:href')
+  t.equal(result.childNodes[2].getAttribute('xlink:href'), '#test', 'created child use tag with xlink:href')
```
Comment (yoshuawuyts) [inline test/elements.js]: think we're close, but so far the goal has been to remove artifacts like these - e.g. when we're creating lists of items that we must reorder, not having any random text nodes in between is a requirement. I should've expressed that as a test tho, my bad - but this test should be changed back

File: test/elements.js [hunk 2]
```diff
@@ -104,6 +104,53 @@ test('adjacent text nodes', function (t) {
...
 
+test('space in only-child text nodes', function (t) {
+  t.plan(1)
+  var result = bel`
+    <span>
+      surrounding
+      newlines
+    </span>
+  `
+  t.equal(result.outerHTML, '<span>surrounding\n      newlines</span>', 'should remove extra space')
```
Comment (yoshuawuyts) [inline test/elements.js]: Perhaps my intuition of how HTML works is wrong, but wouldn't this be closer to what folks might expect:
```js
 t.equal(result.outerHTML, '<span>surrounding newlines</span>', 'should remove extra space')
```

File: test/elements.js [hunk 3]
```diff
@@ -104,6 +104,53 @@ test('adjacent text nodes', function (t) {
...
+
+test('space between non-text nodes', function (t) {
+  t.plan(1)
+  var result = bel`
+    <div>
+      <dfn>whitespace</dfn>
+      <em>is beautiful</em>
+    </div>
+  `
+  t.equal(result.outerHTML, '<div><dfn>whitespace</dfn> <em>is beautiful</em></div>', 'should have correct output')
```
Comment (yoshuawuyts) [inline test/elements.js]: perhaps if we apply the heuristic that if "inline-text nodes" follow each other (or an actual text node), we add a space. E.g.
```html
<strong>hi</strong>
<em>hello</em>
<span>world</span>
planet
<div>bar</div>
```
```html
<strong>hi</strong> <em>hello</em> <span>world</span> planet<div>bar</div>
```

I don't think we can account for all edge cases (html is huuge); but as long as common patterns are supported then we might get away with it (:

File: test/elements.js (hunk 1 above)
Comment (kemitchell) [inline test/elements.js]: That's the tricky case that I tried to address this in the last paragraph of the PR.

The trouble is differentiating, say:

```js
bel`
  <svg width="150" height="100" viewBox="0 0 3 2">
    <rect width="1" height="2" x="0" fill="#008d46" />
    <use xlink:href="#test" />
  </svg>
`
```

from:

```js
bel`
  <p>
    <strong>Goal</strong>
    <em>Do the right thing.</em>
  </p>
`
```

As a practical matter, any HTML element can end up with `white-space: preserve`, and need space as it was written.  This PR tries to address the most common case, `<pre>`.  Unless we start using other tag names to intelligently guess when space between elements should be removed, we have to make a trade-off that affects both of my examples above.

File: test/elements.js (hunk 2 above)
Comment (kemitchell) [inline test/elements.js]: An earlier version of this patch collapsed space within text nodes, as well.

Other than the negligible RAM for the space characters, is there any real perf benefit to messing with text nodes that will definitely end up in the rendered tree?

**Comment by kemitchell** [comment]
@yoshuawuyts Can you give me an example of "keyed reordering of nodes"?  I want to make sure I understand.  I'm guessing it's a diff-algo heuristic?

Two thoughts:

1.  If the optimization is peculiar to a diff algo, isn't there an argument for making bel as generic as possible, and applying a uniform transform---say, that removes known-pointless text nodes---to bel's output before passing it to diff?  That way bel stays simple and clean, and doesn't hide any subtle deviations from well ingrained HTML expectations.

2.  I alluded to, but didn't flesh out, another approach that I like more and more: put the choice to optimize in the API and let users opt into it selectively.

For example:

```js
var bel = require('bel')
var paragraph = bel`
  <p>
    Click
    <a href=/help>here</a>
    for help.
  </p>
`
var render = bel.collapse`
  <header>Site Map</header>
  <main>
    <aside>
      ${paragraph}
    </aside>
  </main>
`
```

would render:

```html
<header>Site Map</header><main><aside><p>
    Click
    <a href=/help>here</a>
    for help.
  </aside></main>
```

File: test/elements.js (hunk 2 above)
Comment (yoshuawuyts) [inline test/elements.js]: It's not so much about performance in this case, as it is about expectations.

This is what people write:
```html
<span>
  surrounding
  newlines
</span>
```

This is what people see rendered on the DOM:
```html
<span>surrounding newlines</span>
```

Now if people inspect the string content:
```js
// this matches what they see on the screen, and is probably wha
//  they expect the content to look like
var str = 'surrounding newlines'

// this is the actual representation - internals of how strings were parsed leak
var str = 'surrounding\n      newlines'
```

Tho it also goes worth noting strings can't be minified, so for complete applications the random empty spaces tend to add up.

**Comment by yoshuawuyts** [comment]
> @yoshuawuyts Can you give me an example of "keyed reordering of nodes"? I want to make sure I understand. I'm guessing it's a diff-algo heuristic?

@kemitchell https://github.com/yoshuawuyts/nanomorph/issues/8

> If the optimization is peculiar to a diff algo, isn't there an argument for making bel as generic as possible, and applying a uniform transform---say, that removes known-pointless text nodes---to bel's output before passing it to diff? That way bel stays simple and clean, and doesn't hide any subtle deviations from well ingrained HTML expectations.

I think my general feeling is that I'd like bel's output to be as unexpected as possible. E.g. keep the output as close as possible to what DOM node creation through `node.innerHTML = ''` looks like:
screen shot 2017-05-20 at 20 33 41

> I alluded to, but didn't flesh out, another approach that I like more and more: put the choice to optimize in the API and let users opt into it selectively.

I'm confident we can get the algorithm good enough that we don't need to implement multiple modes :sparkles:

File: test/elements.js (hunk 1 above)
Comment (yoshuawuyts) [inline test/elements.js]: > As a practical matter, any HTML element can end up with white-space: preserve

Good point, didn't consider that. I wonder how common that is tho, and if that couldn't replaced by using a `<pre>` tag instead (unless we're talking about the most edgy of edge cases.

> Unless we start using other tag names to intelligently guess when space between elements should be removed

I think indeed that including these tags might indeed be the way to go

**Comment by kemitchell** [comment]
@yoshuawuyts Thanks for the links.

IIRC, `innerHTML` is de-facto, but not standardized.  I'm not sure whether it applies the browser's CSS logic to differential block formatting context from inline formatting context or not.  For reference, from CSS3 WD:

https://www.w3.org/TR/css-text-3/#white-space-property

https://www.w3.org/TR/css-text-3/#white-space-processing

**Comment by yoshuawuyts** [comment]
> IIRC, innerHTML is de-facto, but not standardized

Haha, didn't know that - hadn't seen the specs before. But isn't it so that people expect certain space behavior, even if it might not be fixed in a spec? Wouldn't it make sense to implement the behavior people generally expect?

Or am I missing something? Hope we're not talking past each other haha

**Comment by kemitchell** [comment]
> Or am I missing something? Hope we're not talking past each other haha

@yoshuawuyts, brother, we're cool!  Don't sweat it.  And big thanks for speaking my language ;-)

On `innerHTML`, it looks like I'm partially out of date. `innerHTML` and `outerHTML` are in the current HTML5 working draft. The draft spec says setting `innerHTML` invokes the standard HTML5 parser.

It's been a while since I was deep down in the specs, but I believe HTML itself pretty much leaves whitespace alone. There are some rare exceptions, like the first newline in a `<pre>` tag.  But it's the styling engine that does most of the collapsing and trimming.

Oddly, enough, I just fired up Chrome 58.0.3029.81 to try your console example. It looks like the parser---if not the pretty printer---retains pretty much all space:

```
> var div = document.createElement('DIV')
undefined
> div.innerHTML = '<div>\n<div>hello</div>\n<div>hello</div>\n</div>'
"<div>
<div>hello</div>
<div>hello</div>
</div>"
> div
<div>
  <div>
    <div>hello</div>
    <div>hello</div>
  </div>
</div>
> div.childNodes[0]
<div>​…​</div>​
> div.childNodes[0].childNodes
(5) [text, div, text, div, text]
> div.childNodes[0].children
(2) [div, div]
> div.childNodes[0].childNodes[0].data
"
"
```

Note that `children` returns only `Element` child nodes, and that the Chrome pretty printer automatically indents trees.

**Comment by greghuc** [comment]
Just wanted to add support for this whitespace patch, or equivalent.

I upgraded to bel 5.0.0, and was hit by some whitespace surprises (see example below). These whitespace changes have been unsettling as, for the first time, what I express as HTML to bel (as a tagged template string) isn't guaranteed to be accurately reflected in the DOM tree rendered by the browser. So bel is no longer an invisible 'just works' library: developers will need an understanding of bel's internal whitespace-collapsing behaviour in order to correctly supply HTML input. I'm hoping this is a bug, not a feature :-/

Most recent breaking whitespace example was a form, where spacing vanished between a radio button and its label:

```
4.6.0 bel output:

<form>
   <div>
      <input type="radio">
      <label>Create an account</label>
   </div>
</form>

Renders in Chrome as..

* Create an account

5.0.0 bel output:

<form>
   <div>
      <input type="radio"><label>Create an account.</label>
   </div>
</form>

Renders in Chrome as (missing whitespace)..

*Create an account
```

**Comment by yoshuawuyts** [comment]
@greghuc thanks for expressing your concerns. If you're experiencing any issues I can recommend sticking to 4.0.0 for now.

Regarding whitespace though; I'd argue that the previous behavior was rather unpredictable:
```html

<div>
  <p>hello world</p>
</div>

```

How many text nodes will there be generated? 1? 2? Or how about this example:

```html
<div>
  <p>
    hello world
  </p>
</div>
```

Are there more text nodes now? If so: where? How many? What's the right `childNode` to select to get the `<p>` tag? It's tricky, right?

The short answer to this question is: for many people this behavior is rather unpredictable. Libraries like JSX realize this, and they remove all unintended text nodes. Not only does this deliver a cleaner, more predictable DOM - it also helps boost performance as it reduces the DOM node count by large amounts - this means up to !2x improvement on diffing speeds, among other things.

As pointed out, everything has tradeoffs. In our case it'll change the behavior compared to we need to do some work on making sure we preserve whitespace between the right nodes. This takes a little bit of work, and indeed changes the behavior - but it delivers a more robust experience that we can all rely upon. I hope this makes sense. Thanks!

---
P.s. In bel 4 the answer to the first question was __4__, the answer to the second was __2__. The first example has trailing and leading newlines that create text nodes. You might not see these nodes from the markup, but you still need to account for these (e.g. yay, unexpected whitespace behavior). Once this patch lands the answer to both would become __1__: just the content of the `<p>` tag, kinda like what you'd expect.

**Comment by greghuc** [comment]
@yoshuawuyts thanks for the response re whitespace.

Once whitespace behaviour is locked in, then a short explanation in the README would be great, including how to "force whitespace" if needed. I'd hope 'forcing' can be done in the HTML string (i.e. in the tagged template string itself), rather than using `${document.createTextNode(' ')}`. Or if everything just works as expected, need no forcing explanation is needed.

Regarding 'predictability', it depends on who the target audience is. For library builders (and maybe some developers), a DOM without 'unintended text nodes' is no doubt predictable. But for developers/designers writing straight HTML - who aren't focussed on internal details of rendering libraries - then predictable means "bel HTML renders visually in the browser as if I'd loaded the HTML straight from a static .html file".

I hope there's a definition of 'predictable whitespace' for bel that works for both groups :-)

**Comment by yoshuawuyts** [comment]
ping @kemitchell - are you planning to work on this patch further? Is there any way I can assist? - if not let me know; this is an important issue to fix and we shouldn't sit on it for too long haha. Thanks!

**Comment by kemitchell** [comment]
@yoshuawuyts Checking back in. For some reason, I stopped getting e-mail notifications!

From my point of view, the answers to "guess how many text nodes" were clear.  The counts output by the parser spec'd in the standard are the right answers.  More practically, you can load the markup in a web browser and traverse with JS.

As I understand it, the 4.x to 5.x change was performance-motivated, through and through. It's expensive to diff and patch a bunch of text nodes, and there are good, if imperfect, heuristics for deciding when space matters. So it's worth some thought.

But to your point, HTML is huge.  We're unlikely to write a perfect algorithm based on `nodeName`. In fact, that wouldn't be enough. CSS can change the presentation of various elements, say with `display: inline` or `display: inline-block`.  So the best we could do is a "should space stay or should it go" algorithm that's mostly good enough, and document it.

Insofar as all of this is really about optimization, and the good choice right now boils down to pinning 4.x to preserve too much space and rolling with 5.x to preserve too little, I like the idea of exposing the optimization in the API more and more.  I'm going to end up implementing a stripped down version of that API on top of hyperx for one particular app, at least, because the perf gains are just that good, one the one hand, and the purpose of the app is to display pages and pages of highly structured text and inline elements, on the other.  In a way, it reminds me of the various memoization optimizations for tree generation.  They usually don't come standard.  If you hit a perf wall or your app stabilizes, then you go in and apply to the hot paths, as an upgrade.

@yoshuawuyts, I am so super grateful for your work here on bel, not to mention all the other places!  If you want to try to do the best you can with `nodeName` or similar, feel free to start with my code if that's helpful.  But I won't lie and say I have it in me to turn this patch into that kind of rule.  It's an awfully big project for a little module!
